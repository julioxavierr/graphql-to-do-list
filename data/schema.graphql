"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""The root of all... queries"""
type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID
  ): Node

  """Fetches objects given their IDs"""
  nodes(
    """The IDs of objects"""
    ids: [ID!]!
  ): [Node]!
  tasks: TaskConnection
}

"""Represents a Task"""
type Task implements Node {
  """The ID of an object"""
  id: ID!

  """Description of this task"""
  description: String!

  """Wheter this task is checked/unchecked"""
  checked: Boolean!

  """date of creation"""
  createdAt: String!

  """date of last update"""
  updatedAt: String!
}

"""A connection to a list of items."""
type TaskConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [TaskEdge]
}

"""An edge in a connection."""
type TaskEdge {
  """The item at the end of the edge"""
  node: Task

  """A cursor for use in pagination"""
  cursor: String!
}
