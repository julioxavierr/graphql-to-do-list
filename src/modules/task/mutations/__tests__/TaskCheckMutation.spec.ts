import { graphql } from 'graphql';
import idx from 'idx';

import { toGlobalId } from 'graphql-relay';
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  getContext,
  createTask,
} from '../../../../../test/helper';
import { schema } from '../../../../schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should check a task', async () => {
  const task = await createTask();

  // language=GraphQL
  const query = `
    mutation M($task: ID!, $checked: Boolean!) {
      TaskCheck(input: {
        task: $task
        checked: $checked
      }) {
        error
        task {
          description
          checked
        }
      }
    }
  `;

  const rootValue = {};
  const context = await getContext({});
  const variables = {
    task: toGlobalId('Task', task._id),
    checked: true,
  };

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
  expect(idx(result, _ => _.data.TaskCheck.task.checked)).toBe(true);
});

it('should uncheck a task', async () => {
  const task = await createTask({ checked: true });

  // language=GraphQL
  const query = `
    mutation M($task: ID!, $checked: Boolean!) {
      TaskCheck(input: {
        task: $task
        checked: $checked
      }) {
        error
        task {
          description
          checked
        }
      }
    }
  `;

  const rootValue = {};
  const context = await getContext({});
  const variables = {
    task: toGlobalId('Task', task._id),
    checked: false,
  };

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
  expect(idx(result, _ => _.data.TaskCheck.task.checked)).toBe(false);
});
