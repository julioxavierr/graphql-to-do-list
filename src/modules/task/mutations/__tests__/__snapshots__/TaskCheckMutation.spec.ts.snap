// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should check a task 1`] = `
Object {
  "data": Object {
    "TaskCheck": Object {
      "error": null,
      "task": Object {
        "checked": true,
        "description": "task_1",
      },
    },
  },
}
`;

exports[`should uncheck a task 1`] = `
Object {
  "data": Object {
    "TaskCheck": Object {
      "error": null,
      "task": Object {
        "checked": false,
        "description": "task_1",
      },
    },
  },
}
`;
